File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-28.js
// class Node {
//   constructor(val, next = null, random = null) {
//       this.val = val;
//       this.next = next;
//       this.random = random;
//   }
// }

class Solution {
    /**
     * @param {Node} head
     * @return {Node}
     */
    copyRandomList(head) {  
        if (!head) return null
        let map = new Map() 

        let node = head 
        while(node) {
            const copy = new Node(node.val) 
            map.set(node, copy) 
            node = node.next
        } 

        node = head 
        while(node) { 
            let copy = map.get(node) 
            let next = node.next 
            let random = node.random 

            copy.next = map.get(next) || null 
            copy.random = map.get(random) || null
            node = node.next
        } 

        return map.get(head)
    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-28.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head) 
        groupPrev = dummy 

        while True: 
            start = groupPrev
            kNode = self.getK(start, k) 
            if not kNode:
                break

            start = groupPrev.next
            groupNext = kNode.next  
            prev = groupNext 

            while start != groupNext: 
                temp = start.next 
                start.next = prev 
                prev = start 
                start = temp 

            temp = groupPrev.next 
            groupPrev.next = prev 
            groupPrev = temp 

        return dummy.next

    def getK(self, node, k): 

        while k > 0 and node: 
            node = node.next 
            k -= 1 

        return node